refactor: Extract statement construction functions, use `async`/`await`, begin extracting tests.

emmanuel opened the pull request:
Main thing here is extracting statement construction from other logic. 

First goal is to support simpler and more extensive testing of statement construction (see `cmi5-statements.spec.ts`), but secondarily it opens up the library to additional (unforeseen) use-cases by providing CMI5-compliant xAPI statement construction ala carte from the LRS client. If you're into patterns, this PR could be described as an application of the "single responsibility principle" pattern (statement construction vs LRS communication).

Stylistically, this de-emphasizes classes and focuses on functions and data types.

emmanuel committed:
refactor: Use async/await & add XAPI statement constructors.

emmanuel committed:
refactor: Separate xAPI statement construction from API calls.

emmanuel committed:
refactor: Extract helpers, begin extracting statement tests.

Much of the variation in testing can be handled by validating statement construction. Almost all of it, really, with the notable exceptions of `initialize()` and `moveOn()`.

Anyhow, I intend to extract the remaining statement tests from the `Cmi5` class tests (`cmi5.spec.ts`), and to extend the statement tests to cover more cases.

emmanuel committed:
cleanup: Separate `Context...`/`ResultExtension`, run `prettier`.

CookieCookson commented:
@emmanuel This is a really helpful contribution, thank you! If you get it rebased I can give it a more thorough review next week, at first glance it is a nice refactor and test addition.

emmanuel commented:
@CookieCookson Great! I'm glad this looks like a useful contribution. And thanks for the encouragement!

I rebased and tests still pass locally.

emmanuel commented:
Actually, regarding the tests: I dropped the mocking of time in the statement tests, which mostly works, but occasionally fails. IOW, I introduced a flaky test. 

I have a thought about how to address that (my intention is to do so without returning to mocking), but it's not implemented here. Right now, if you get a test failure on the duration tests like this:
```
    Expected: "PT2H6M33S"
    Received: "PT2H6M33.001S"
```
That's the flake. Re-run the tests (maybe more than once) and it will pass. I've only seen the failure locally (not in CI), but it _could_ happen in CI, too.

emmanuel commented:
Hi @CookieCookson, how's it going? 

Have you had a chance to review this? Is there anything I can do to help?

CookieCookson commented:
@emmanuel I've not had any time again, sorry. I will see if I can get some time this week to investigate the flaky test(s). I can see it just being from slow execution time or a rounding error.

emmanuel commented:
To be clear, I wasn’t expecting you to investigate the test I introduced! I imagined merging the flaky test and hitting retry on CI if it fails there. Not the cleanest, perhaps, but addressing it properly may be a significant step up in complexity. That said, I didn’t realize that’s what you’re waiting on. If so, let me take another pass and resolve the intermittent fail on the XAPI duration. 

emmanuel committed:
fix: Resolve flaky test without new dependencies.

emmanuel commented:
@CookieCookson alright, I resolved the flake. 

In lieu of mocking, stubbing, or adding a dependency (e.g., `.toBeOneOf()` from https://github.com/jest-community/jest-extended), I took a ... creative ... approach. Viz., I flipped actual and expected in order to leverage the built-in `.toContain()` matcher 😁.

emmanuel commented:
I'd be delighted to see this land soon. 

Please let me know if there's anything else you'd like to see before merging this.

CookieCookson commented:
Excellent and creative resolution - great job! I should have time this weekend to get this reviewed and released.

Did you see my comment on the other closed PR? I'm concerned about how you may actually consume the changes made as it doesn't seem that anything new is being exported.

emmanuel commented:
@CookieCookson thanks for looking again. 

I looked back at the last PR ("support Cmi5 from the backend"), and I think you're referring to [this comment](https://github.com/xapijs/cmi5/pull/266#issuecomment-2321126209), and yet ...I'm confused 🤔. 

The `AbstractCmi5` class [is exported as the default for the file in the original PR](https://github.com/xapijs/cmi5/pull/266/files#diff-23c43be2b1773ba2786d6fd8df00a912133f972dbbda6532951cd6f492a8a345R80), and [that remains the case in this PR](https://github.com/xapijs/cmi5/pull/274/files#diff-23c43be2b1773ba2786d6fd8df00a912133f972dbbda6532951cd6f492a8a345R61). Also, all of the statement creation functions are exported.

So... to what were you referring?

CookieCookson commented:
I was thinking that with needing to use this on the backend, you will need to install the package and import the Abstract CMI class. When this package builds, it uses the original Cmi5 class as the entry point so you will be forced to use the class that uses the browser APIs which you may be trying to avoid. I may be confused here too, it would be good to have a code example of how you plan on consuming the Abstract CMI class in another project.

CookieCookson commented:
@emmanuel I finally found some time to give this the time it needs! The code change was substantial so it took a fair while to review but its lots of good stuff. All merged, this plus your previous contributions plus some security vulnerability patches have been released into `@xapi/cmi5@1.4.0`.

A huge thank you for your contributions on this project, and an apology that its taken months and months to get these code changes out into the wild.

Is https://github.com/xapijs/cmi5/issues/265 now ready to be closed up or are there additional code changes still required for this?

emmanuel commented:
I went ahead and close #265; all done there!

Thanks for your encouragement and review. And again, thanks for publishing this project.
